server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -485,4 +544,163 @@ private static String signatureMessage(final Class<? extends Plugin> clazz) {
     public <T> List<T> filterPlugins(Class<T> type) {
         return plugins.stream().filter(x -> type.isAssignableFrom(x.instance().getClass())).map(p -> ((T) p.instance())).toList();
     }
+
+    static final LayerAndLoader createPluginModuleLayer(PluginBundle bundle, ClassLoader parentLoader, List<ModuleLayer> parentLayers) {
+        assert bundle.plugin.getModuleName().isPresent();
+        return createModuleLayer(
+            bundle.plugin.getClassname(),
+            bundle.plugin.getModuleName().get(),
+            urlsToPaths(bundle.urls),
+            parentLoader,
+            parentLayers
+        );
+    }
+
+    static final LayerAndLoader createSpiModuleLayer(
+        String moduleName,
+        Set<URL> urls,
+        ClassLoader parentLoader,
+        List<ModuleLayer> parentLayers
+    ) {
+        // assert bundle.plugin.getModuleName().isPresent();
+        return createModuleLayer(
+            null,  // no entry point
+            moduleName + ".spi",  // by convention ?
+            urlsToPaths(urls),
+            parentLoader,
+            parentLayers
+        );
+    }
+
+    private static final Module serverModule = PluginsService.class.getModule();
+
+    static final LayerAndLoader createModuleLayer(
+        String className,
+        String moduleName,
+        Path[] paths,
+        ClassLoader parentLoader,
+        List<ModuleLayer> parentLayers
+    ) {
+        logger.debug(() -> "creating module layer and loader for module " + moduleName);
+        var finder = ModuleFinder.of(paths);
+
+        // TODO resolve AND BIND ?
```

Comment:
is this todo clearable or can you expand the comment so it is clear what still needs to happen?

Reply:
resolveAndBind is better. Updated.